### PR TITLE
Add major/minor axis length and ratio

### PR DESCRIPTION
**Pull request by jsta**
See #4

### CODE REVIEW

File: R/lakeMajorAxisLength.R [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  # points(elpshull.center, col = "red")
+  # lines(rbind(elpshull.center, elpshull[dist2center == min(dist2center),]))
+  # lines(elpshull[dist2center == max(dist2center),])
+  
+  # myLine.min <- rbind(elpshull.center, elpshull[dist2center == min(dist2center),])
+  myLine.max <- elpshull[dist2center == max(dist2center),]
+  myLine <- SpatialLines(list(Lines(list(Line(myLine.max)), "1"))
+, proj4string = CRS(proj4string(inLakeMorpho$lake)))
+
+  result <- max(dist2center)
```
Comment (jhollist) [inline R/lakeMajorAxisLength.R]: Is this identical to the length of the line that represents the major axis.  Not up enough on my ellipsoids to know, but it looks like it should be...  If not, we should use gLength on myLine instead.

File: R/lakeMajorAxisLength.R [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+lakeMajorAxisLength <- function(inLakeMorpho, pointDens, addLine = TRUE) {
+  if (class(inLakeMorpho) != "lakeMorpho") {
+    stop("Input data is not of class 'lakeMorpho'.  Run lakeSurround Topo or lakeMorphoClass first.")
+    stop("Input data is not of class 'lakeMorpho'.  Run lakeSurround Topo or lakeMorphoClass first.")
+  }
+  
+  result <- NA
+  lakeShorePoints <- spsample(as(inLakeMorpho$lake, "SpatialLines"), pointDens, "regular")@coords
+  # chull <- gConvexHull(lakeShorePoints)
```
Comment (jhollist) [inline R/lakeMajorAxisLength.R]: Clean up the excess comments

File: R/lakeMajorAxisLength.R [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  
+  # https://stackoverflow.com/questions/18278382/how-to-obtain-the-lengths-of-semi-axes-of-an-ellipse-in-r
+  elpshull <- predict(cluster::ellipsoidhull(lakeShorePoints))
```
Comment (jhollist) [inline R/lakeMajorAxisLength.R:40]: Is there a significant slow down on building the hull with a greater number of points?  Reason I ask is that I am thinking about redoing the pointdens stuff (see #6) to default to all points along the polygon bnd with the point density selecting a proportion of those points to allow for speeding up lakeMaxLength etc.  If the ellipsoid prediction is not significantly impacted by the number of points, you could just default to the full set of points via as(as(lake,"SpatialLines"),"SpatialPoints") or similar.

File: R/lakeMinorAxisLength.R [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  dist2center <- sqrt(rowSums((t(t(elpshull) - colMeans((elpshull))))^2))
+  # max(dist2center)     ## major axis
+  # min(dist2center)   ## minor axis
+
+  myLine.min <- rbind(elpshull.center, elpshull[dist2center == min(dist2center),])
+  
+  myLine <- SpatialLines(list(Lines(list(Line(myLine.min)), "1")),
+              proj4string = CRS(proj4string(inLakeMorpho$lake)))
+
+  result <- max(dist2center)
```
Comment (jhollist) [inline R/lakeMinorAxisLength.R]: should this be min(dist2center)??

File: R/lakeMinorAxisLength.R [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  }
+  
+  result <- NA
+  lakeShorePoints <- spsample(as(inLakeMorpho$lake, "SpatialLines"), pointDens, "regular")@coords
+  # chull <- gConvexHull(lakeShorePoints)
+  
+  # https://stackoverflow.com/questions/18278382/how-to-obtain-the-lengths-of-semi-axes-of-an-ellipse-in-r
+  elpshull <- predict(cluster::ellipsoidhull(lakeShorePoints))
+  elpshull.center <- matrix(colMeans((elpshull)), ncol = 2, nrow = 1)
+
```
Comment (jhollist) [inline R/lakeMinorAxisLength.R:37]: Back to the timing on this.  If it is quick we can leave as is, but if it takes a while for larger polys then perhaps we could store both the major/minor axis and the ellipsoid on the lakeMorpho object, thus if it has been run for either major or minor already, we don't have to calculate it again.

File: R/lakeMajorAxisLength.R [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  
+  dist2center <- sqrt(rowSums((t(t(elpshull) - colMeans((elpshull))))^2))
+  # max(dist2center)     ## major axis
+  # min(dist2center)   ## minor axis
+
+  myLine.max <- elpshull[dist2center == max(dist2center),]
+  myLine <- SpatialLines(list(Lines(list(Line(myLine.max)), "1")),
+              proj4string = CRS(proj4string(inLakeMorpho$lake)))
+
+  result <- max(dist2center)
```
Comment (jhollist) [inline R/lakeMajorAxisLength.R]: Is this different that result we'd get from gLength(myLine)?  Not up enough on my ellipsoids/convex hulls to know.  All the other results use gLength to pull this value.

File: R/lakeMinorMajorAxisRatio.R [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+#'        in practice is usually significant less.
+#' @param addLine Boolean to determine if the selected major and minor axis lines
+#'        should be added to the inLakeMorpho object.  Defaults to True
+#' 
+#' @export
+#' @return This returns a vector of numeric values indicating the length of the
+#'         major and minor axes of the lake. Units are the same as the input data.
+#' 
+#' @references \href{https://en.wikipedia.org/wiki/Semi-major_and_semi-minor_axes}{Link}
+#' @import sp rgeos methods
```
Comment (jhollist) [inline R/lakeMinorMajorAxisRatio.R]: I think we'd want to either use the pkg::function, @import or @importFrom.    Also not sure we need to have all these imports for this function.  Looks like just rgeos is needed.  My current preference is to specify with pkg::function, but I think I used @import in the rest of `lakemorpho`.  One or the other is fine with me.

**Review by jhollist — COMMENTED** [review]
@jsta I think these will be a very nice addition. Couple of general issues

- Do you have limnology examples or citations of major and minor axis?  If so, be nice to add those citations into the function documentation
- `cluster` needs to be added to Imports in the DESCRIPTION
- Be sure to add yourself, with role "ctb" in Authors@R in the DESCRIPTION

Not sure when I will get around to rolling this all up as a new CRAN release.  Hopefully not later than end of Jan.

Thanks again for the contribution!

**Comment by jsta** [comment]
Ok, I think I addressed all of your comments with the exception of the limnology references. I know minor-major axis ratio is used in image classification analysis but I am not aware of specific limnology references.

**Comment by jhollist** [comment]
@jsta 

Looks great.  Merging shortly.  I *may* get around to pushing the update to CRAN in the next day or so.

Thanks again for making the contribution!

**Comment by jhollist** [comment]
FYI @jsta, just submitted to CRAN!

**Pull request merged**